sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala

```diff
@@ -134,8 +142,26 @@ object FileFormatWriter extends Logging {
       // prepares the job, any exception thrown from here shouldn't cause abortJob() to be called.
       committer.setupJob(job)
 
+      val bucketIdExpression = bucketSpec.map { spec =>
+        // Use `HashPartitioning.partitionIdExpression` as our bucket id expression, so that we can
+        // guarantee the data distribution is same between shuffle and bucketed data source, which
+        // enables us to only shuffle one side when join a bucketed table and a normal one.
+        HashPartitioning(bucketColumns, spec.numBuckets).partitionIdExpression
+      }
+      // We should first sort by partition columns, then bucket id, and finally sorting columns.
+      val requiredOrdering = (partitionColumns ++ bucketIdExpression ++ sortColumns)
+        .map(SortOrder(_, Ascending))
+      val actualOrdering = queryExecution.executedPlan.outputOrdering
+      // We can still avoid the sort if the required ordering is [partCol] and the actual ordering
+      // is [partCol, anotherCol].
+      val rdd = if (requiredOrdering == actualOrdering.take(requiredOrdering.length)) {
```

Comment:
We only care if partition columns are the same between requiredOrdering and actualOrdering. The sort direction doesn't matter.